### ISSUE TITLE
Consider flattening anonymous structs/unions into the parent struct.

### ISSUE THREAD

**Issue by emilio** [servo/rust-bindgen#784]
As requested by @RReverser in #732.

This is slightly involved and requires a fair amount of tests, but should be doable.

Note the comments in #732 for some potential pitfalls. Implementing this would require to generate padding fields too.

**Comment by Hugal31** [comment]
Tell me if I am wrong, but I'm not sure if we can flatten anonymous union in Rust.

For instance, if

```c
struct Foo {
    union {
        uint64_t word;
        uint32_t bar;
    };
    uint32_t baz;
};
```
becomes

```rust
#[repr(C)]
struct Foo {
    word: u64,
    bar: u32,
    baz: u32,
}
```

how would you tell rust that word and bar are at the same offset ?

**Comment by RReverser** [comment]
@Hugal31 Word and bar would likely still be generated as a union here and use getters on the `Foo`, but this is more relevant to #732.

My original request was just about structures, which are more straightforward, I believe @emilio added unions to this issue for completeness (or maybe by mistake).

**Comment by emilio** [comment]
You can in theory have nested anonymous unions I guess, which you could theoretically flatten. But yeah you can't flat structs with unions or vice versa.

**Comment by RReverser** [comment]
@emilio Hmm, how would you flatten nested unions?

```c
union Foo {
    union {
        uint64_t word;
        uint32_t bar;
    };
    uint32_t baz;
};
```

This would still require getters, no?

**Comment by emilio** [comment]
Yes, both getters and padding I guess, but it's the same issue with structs, right? Unless I'm misunderstanding your question.

In that case the storage for the union could be a u64, and word would read it, and bar and baz would read the same memory.

**Comment by RReverser** [comment]
For just nested structs optional padding should be always enough and getters won't be required, while for unions I don't see how to get direct flattening without getters.
